// appcenter-file-upload-client-node/out/src/ac-fus-uploader-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ACFusMessageLevel;
(function (ACFusMessageLevel) {
    ACFusMessageLevel[ACFusMessageLevel["Information"] = 0] = "Information";
    ACFusMessageLevel[ACFusMessageLevel["Verbose"] = 1] = "Verbose";
    ACFusMessageLevel[ACFusMessageLevel["Error"] = 2] = "Error";
})(ACFusMessageLevel = exports.ACFusMessageLevel || (exports.ACFusMessageLevel = {}));
var ACFusUploadState;
(function (ACFusUploadState) {
    ACFusUploadState[ACFusUploadState["New"] = 0] = "New";
    ACFusUploadState[ACFusUploadState["Initialized"] = 10] = "Initialized";
    ACFusUploadState[ACFusUploadState["Uploading"] = 20] = "Uploading";
    ACFusUploadState[ACFusUploadState["ResumeOrRestart"] = 40] = "ResumeOrRestart";
    ACFusUploadState[ACFusUploadState["Paused"] = 50] = "Paused";
    ACFusUploadState[ACFusUploadState["Error"] = 60] = "Error";
    ACFusUploadState[ACFusUploadState["Cancelled"] = 80] = "Cancelled";
    ACFusUploadState[ACFusUploadState["Verifying"] = 90] = "Verifying";
    ACFusUploadState[ACFusUploadState["Completed"] = 100] = "Completed";
    ACFusUploadState[ACFusUploadState["FatalError"] = 500] = "FatalError";
})(ACFusUploadState = exports.ACFusUploadState || (exports.ACFusUploadState = {}));
